device/daita.go

```diff
@@ -0,0 +1,301 @@
+//go:build daita
+// +build daita
+
+package device
+
+import (
+	"time"
+	"unsafe"
+)
+
+// #include <stdio.h>
+// #include <stdlib.h>
+// #include "../cmaybenot/cmaybenot.h"
+// #cgo LDFLAGS: -L${SRCDIR}/../cmaybenot/target/release -l:libcmaybenot.a -lm
+import "C"
+
+type MaybenotDaita struct {
+	events         chan *Event
+	actions        chan *Action
+	maybenot       *C.Maybenot
+	newActionsBuf  []C.MaybenotAction
+	machineActions map[uint64]Action
+	logger         *Logger
+}
+
+type Event struct {
+	// The machine that generated the action that generated this event, if any.
+	Machine uint64
+
+	Peer      NoisePublicKey
+	EventType EventType
+	XmitBytes uint16
+}
+
+type ActionType uint32
+
+const (
+	Cancel        = ActionType(0)
+	InjectPadding = ActionType(1)
+	BlockOutgoing = ActionType(2)
+)
+
+const (
+	ERROR_GENERAL_FAILURE      = -1
+	ERROR_INTERMITTENT_FAILURE = -2
+)
+
+type EventContext struct {
+	tunnelHandle int32
+	peer         NoisePublicKey
+}
+
+type Action struct {
+	Peer NoisePublicKey
+
+	ActionType ActionType
+
+	// The maybenot machine that generated the action.
+	// Should be propagated back by events generated by this action.
+	Machine uint64
+
+	// The time at which the action should be performed
+	Time time.Time
+
+	// TODO: Support more action types than InjectPadding
+	Payload Padding
+}
+
+type Padding struct {
+	ByteCount uint16
+	Replace   bool
+}
+
+// TODO: Turn off DAITA? Remember to send a nil action when doing so
+func (device *Device) EnableDaita(machines string, eventsCapacity uint, actionsCapacity uint) bool {
+	if device.Daita != nil {
+		device.log.Errorf("Failed to activate DAITA as it is already active")
+		return false
+	}
+
+	device.log.Verbosef("DAITA activated")
+	device.log.Verbosef("Params: eventsCapacity=%v, actionsCapacity=%v", eventsCapacity, actionsCapacity) // TODO: Deleteme
+
+	var maybenot *C.Maybenot
+	c_machines := C.CString(machines)
+	maybenot_result := C.maybenot_start(
+		(*C.int8_t)(c_machines), 0.0, 0.0, 1440,
```

Comment:
You probably want to determine the actual MTU being used on the device, and/or allow this to be configured.

Reply:
Yep